File: 0064-minimum-path-sum/0064-minimum-path-sum.cpp
class Solution {
public:
    int f(vector<vector<int>>& grid, int i, int j, vector<vector<int>>& dp){
        if(i==grid.size()-1 && j==grid[0].size()-1){
            return dp[i][j]=grid[grid.size()-1][grid[0].size()-1];
        }
        if(i>=grid.size() || j>=grid[0].size()){
            return INT_MAX;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        return dp[i][j]=min(f(grid,i+1,j,dp),f(grid,i,j+1,dp))+grid[i][j];
    }
    int minPathSum(vector<vector<int>>& grid) {
        int m=grid.size();
        int n=grid[0].size();
        vector<vector<int>>dp(m+1,vector<int>(n+1,-1));
        return f(grid,0,0,dp);
    }
};

File: 0064-minimum-path-sum/0064-minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int[][] dp = new int[grid.length][grid[0].length];
        // return solve(grid, grid.length - 1, grid[0].length - 1);
        return tabulation(grid, grid.length, grid[0].length, dp);
    }

    public int solve(int[][] grid, int m, int n) {
        if (m == 0 && n == 0) {
            return grid[0][0];
        }
        int up = m > 0 ? solve(grid, m - 1, n) : Integer.MAX_VALUE;
        int left = n > 0 ? solve(grid, m, n - 1) : Integer.MAX_VALUE;
        return grid[m][n] + Math.min(up, left);
    }

    public int tabulation(int[][] grid, int m, int n, int[][] dp) {
        dp[0][0] = grid[0][0];
        int up = Integer.MAX_VALUE;
        int left = Integer.MAX_VALUE;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 && j == 0) {
                    continue;
                }
                up = i > 0 ? dp[i - 1][j] : Integer.MAX_VALUE;
                left = j > 0 ? dp[i][j - 1] : Integer.MAX_VALUE;
                dp[i][j] = grid[i][j] + Math.min(up, left);
            }
        }
        return dp[m - 1][n - 1];
    }
}